Issue #4741 - fix issues with HttpServletMapping

lachlan-roberts opened the pull request:
**Issue #4741**

There were previously 8 TCK tests which were failing due to `HttpServletMapping`, this PR fixes 7 of them however the remaining one is due to the async dispatch expecting the `HttpServletMapping` of the source and not the target, but we think the spec and tck are wrong on that so it hasn't been fixed for now.

- `HttpServletMapping` is now created and set on the `Request` by the `ServletHandler` and is not created lazily which allows us to get the correct servletName in the mapping.
- Removed the leading / for `matchValues` which is tested for in the TCK.
- For exact matches we now store the `HttpServletMapping` and don't need to re-create it every time `getHttpServletMapping()` is called.

lachlan-roberts committed:
Issue #4741 - fixes to jetty implementation of HttpServletMapping

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4741 - don't lazily generate HttpServletMapping to preserve servletName

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4741 - tests should expect no leading / for matchValue

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
resolving TODOs from review

- removed pathSpec from Request
- getServletMapping moved to ServletHandler

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4741 - only create HttpServletMapping for exact matches once

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
use wrapped attributes for async dispatch

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContextEvent.java:
```diff
@@ -185,4 +183,117 @@ public void addThrowable(Throwable e)
+        static AsyncAttributes fromForwardedAttributes(Attributes attributes)
+        {
+            AsyncAttributes asyncAttributes = new AsyncAttributes(attributes);
+            asyncAttributes._requestURI = (String)attributes.getAttribute(RequestDispatcher.FORWARD_REQUEST_URI);
```
Do we need to get these attributes now?  could we just get them from the wrapped attributes if they are got?

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/AsyncContextEvent.java:
```diff
@@ -185,4 +183,117 @@ public void addThrowable(Throwable e)
+        }
+
+        @Override
+        public void setAttribute(String key, Object value)
```
I'm not entirely sure we need to support setting these attributes?   I think we can let a set of them be a noop.

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -452,17 +455,18 @@ public void doScope(String target, Request baseRequest, HttpServletRequest reque
                 String servletPath = pathSpec.getPathMatch(target);
```
Do we still need to get the pathSpec here?  If so, don't get it twice. get it once before the if.

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -452,17 +455,18 @@ public void doScope(String target, Request baseRequest, HttpServletRequest reque
+                    baseRequest.setHttpServletMapping(httpServletMapping);
                 }
             }
         }
```
in the log debug below, maybe just log the mapping and make sure it has a good toString

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -452,17 +455,18 @@ public void doScope(String target, Request baseRequest, HttpServletRequest reque
+                HttpServletMapping httpServletMapping = mapping.getHttpServletMapping(servletPath);
                 if (DispatcherType.INCLUDE.equals(type))
                 {
                     baseRequest.setAttribute(RequestDispatcher.INCLUDE_SERVLET_PATH, servletPath);
```
It's not really necessary, but lets do the Attributes wrapping trick here as well!  Do we unset these attributes ever?  We should!

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -1731,6 +1739,140 @@ void destroyListener(EventListener listener)
+     */
+    public static class MappedServletResource extends MappedResource<ServletHolder>
+    {
+        private final MappedServlet _mappedServlet;
```
Why is this a field and not just the same class?

gregw reviewed (changes requested):
Looking good! But let's hangout and discuss niggles.

lachlan-roberts committed:
Issue #4741 - Changes from review, revert async attribute wrapping

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@gregw this is ready for re-review

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -555,21 +559,23 @@ public void doHandle(String target, Request baseRequest, HttpServletRequest requ
         if (holder == null)
             return null;
-        return new MappedResource<>(null, holder);
+        return new MappedServlet(null, holder);
```
We probably could return a shared instance for this as well?

gregw replied:
Also invert the if above and finish method with return null

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -1731,6 +1739,109 @@ void destroyListener(EventListener listener)
+                        break;
+                }
+            }
+            else
```
if pathSpec is null then this is a named dispatch so we can probably create a `_httpServletMapping` here as well?

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -434,41 +437,42 @@ public void doScope(String target, Request baseRequest, HttpServletRequest reque
         UserIdentity.Scope oldScope = null;
 
-        MappedResource<ServletHolder> mapping = getMappedServlet(target);
+        MappedServlet mapping = getMappedServlet(target);
```
rename `s/mapping/mapped/` because a mapping is a different class.

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -434,41 +437,42 @@ public void doScope(String target, Request baseRequest, HttpServletRequest reque
+            if (pathSpec != null)
             {
-                PathSpec pathSpec = mapping.getPathSpec();
                 String servletPath = pathSpec.getPathMatch(target);
```
why are we still getting serlvetPath and pathInfo directly from the pathSpec.   I feel like we can do this more efficiently if we asked the mapping, as it will already know if it is an exact match etc.

gregw reviewed (dismissed):
Sorry, it is almost there, but I'd really like to milk this for all we can in efficiency and simplicity.  Perhaps we should pair program to get the last few niggles out?

gregw commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -434,41 +437,42 @@ public void doScope(String target, Request baseRequest, HttpServletRequest reque
                 String servletPath = pathSpec.getPathMatch(target);
                 String pathInfo = pathSpec.getPathInfo(target);
 
+                HttpServletMapping httpServletMapping = mapping.getHttpServletMapping(servletPath);
```
Just to double down on my comments.   The getHttpServletMapping internally does pretty much all the work that is done in getPathMatch and get PathInfo above.   It just needs to remember the results and return them in the JettyHttpServletMapping and we can do only 1 slice&dice rather than 3.

lachlan-roberts committed:
Issue #4741 - Changes from review

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
Issue #4741 Async ServletMapping

Greatly increased the scope of this PR by combining the servletPath and
pathInfo into the ServletPathMapping class that implements the
HttpServletPathMapping interface.    This allows us to greatly simplify
the matching of servlets and reduce the number of times we need to
actually to the match per request.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4741 Async ServletMapping

Fixed problems with previous commit
more cleanup of attributes in dispatcher.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4741 Async ServletMapping

More code cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4741 Async ServletMapping

Named dispatch cleanup

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-4741-getServletMappingAsyncDispatch

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
The changes I've been making here are towards an end where we can use the `ServletPathMapping` everywhere for the `servletPath`.

Currently the request has fields for:
 - requestURI 
 - contextPath
 - servletPath
 - pathInfo
 - servletPathMapping

The invarient is that requestURI = contextPath + servletPath + pathInfo (modulo encoding and null handling)
The request starts with contextPath servletPath set to null and pathInfo==requestURI
Then any ContextHandler strips off the context path, so contextPath is set, servletPath is still null and pathInfo == pathInContext
Then any ServletHandler strips off the servletPath from the pathInfo, so all three are now correctly set and the mapping is also set.

The problems are:
 - that we have now lost pathInContext and we frequently recalculate by addPaths(servletPath, pathInfo)
 - the servletPath and pathInfo are fields in our mapping, so we have duplicated them 
 - to do a dispatch we need to set then reset 3 fields (all 5 for cross context).

What this PR is a step towards is the request only carrying:
 - requestURI 
 - contextPath
 - pathInContext
 - servletPathMapping

To obtain the servletPath is `servletPathMapping==null?null:servletPathMapping.getServletPath()`
To obtain the pathInfo is `servletPathMapping==null?pathInContext:servletPathMapping.getPathInfo()`
This is how all the attribute wrappers are now setup in this PR.   Eventually will do the same for the request.
The advantages are:
 - the pathInContext is directly available and does not need to be calculated
 - no duplication
 - for a dispatch only 1 field needs to be set/reset (3 for cross context).

I decided that changing request in this PR would make it too huge.   So I've just implemented this pattern for attributes and dispatchers.  Request can be done later.

gregw committed:
Issue #4741 Async ServletMapping

misc cleanup

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-4741-getServletMappingAsyncDispatch

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/ServletPathMapping.java:
```diff
@@ -0,0 +1,148 @@
+    private final String _servletPath;
+    private final String _pathInfo;
+
+    public ServletPathMapping(PathSpec pathSpec, String servletName, String pathInContext)
```
Need a comment in here about the purpose of this class, and particularly how it is used to model both mapping via a path, but ALSO mapping via just a name. In fact, I don't like the name of this class containing the word "path" if path has nothing to do with it.

gregw replied:
I would call it just `ServletMapping`, but that name is already taken!  Let me see if we can avoid using it at all for named dispatch....

janbartel replied:
While you're at it, the ServletMapping class could do with a bit of an overhaul:  the setDefault()/isDefaul() method names I now see are confusing with the whole other meaning of "default". In the case of the ServletMapping these methods just mean whether or not the mapping originated from a webdefault.xml. Can you think of a clearer name?

gregw replied:
It turned out that named dispatchers were broken. Added a test for them and now no longer use a ServletPathMapping for named dispatcher.  Added some javadoc as well.

gregw replied:
rename of isDefault done.

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -1731,6 +1689,71 @@ void destroyListener(EventListener listener)
+            }
+            else if (_pathSpec == null)
+            {
+                // Named servlet mapping
```
See my comment on the ServletPathMapping class: this comment at like 1722 is the only clue that the ServletPathMapping class can be used as a name-match only .

gregw replied:
fixed. see above.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/AsyncAttributes.java:
```diff
@@ -94,54 +83,15 @@ public void setAttribute(String key, Object value)
         switch (key)
         {
             case AsyncContext.ASYNC_REQUEST_URI:
```
Seems the intention is to make this immutable. Perhaps we should throw IllegalStateException if setAttribute() is called? That way we would find out where in the code it is being (incorrectly) called, rather than silently failing and being ignored.

Further thoughts: all of the subclasses of Attributes.Wrapper are immutable, see also ProxyAttributes and SslAttributes. Should the Attributes interface be changed to remove setAttribute() and clearAttributes() methods altogether?

gregw replied:
I don't think ISE is right.   I've change to always setting the attribute on the wrapped map, which we will hide during the forward. I added comments to explain.

lachlan-roberts replied:
Why set the attribute on the wrapped map, we never unwrap the AsyncAttributes so this will never be used.

gregw replied:
Current behaviour is to only set the attribute on the wrapped map for non-async attributes.  Those are used by AsyncAttributes as it passes on non-async gets to the the wrapped map.

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/AsyncAttributes.java:
```diff
@@ -94,54 +83,15 @@ public void setAttribute(String key, Object value)
             default:
                 super.setAttribute(key, value);
                 break;
         }
```
Different behaviour here than ForwardAttribute and IncludeAttribute:  both of those do a remove if the value is null, but see comment above about removing setAttribute() entirely.

gregw replied:
Added comment and simplified forward/include.

gregw committed:
Issue #4741 Async HttpServletMapping

Added tests for named dispatchers
Do not use ServletPathMapping for named dispatch

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4741 Async HttpServletMapping

renamed confusing isDefault method on ServletMapping

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4741 Async HttpServletMapping

simplified setAttribute

Signed-off-by: Greg Wilkins <gregw@webtide.com>

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1925,50 +1842,53 @@ public void setAsyncAttributes()
-
         // Unwrap the _attributes to get the base attributes instance.
         Attributes baseAttributes;
         if (_attributes == null)
```
This whole section really needs some comments - the pattern of handling the AsyncAttributes appears to be very different to handling the Forward/IncludeAttributes that we should explain why.

gregw replied:
extra comments in cf0a795475d78d27743bf331b5134a85515c45e7

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -2411,19 +2331,34 @@ else if (oldQueryParams == null || oldQueryParams.size() == 0)
     }
 
-    public void setPathSpec(PathSpec pathSpec)
+    /**
```
These were public methods on Request, should we be removing them in a point release? Maybe we should mark them as Deprecated and just not call them.

joakime replied:
We only used `setPathSpec(PathSpec pathSpec)` on a test case.
There's no practical way to use `Request.setPathSpec(PathSpec)` except in a testcase.

janbartel commented on jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHandler.java:
```diff
@@ -555,29 +523,28 @@ public void doHandle(String target, Request baseRequest, HttpServletRequest requ
      * @param target Path within _context or servlet name
      * @return MappedResource to the ServletHolder.  Named servlets have a null PathSpec
```
Return is wrong: should be returning MappedServlet

gregw committed:
Issue #4741 Async HttpServletMapping

added javadoc about AsyncAttributes

Signed-off-by: Greg Wilkins <gregw@webtide.com>

janbartel reviewed (changes requested):
A few changes listed in the comments.

gregw committed:
Issue #4741 Async HttpServletMapping

Fixed javadoc

Signed-off-by: Greg Wilkins <gregw@webtide.com>

janbartel reviewed (approved):
LGTM
